### DISCUSSION TITLE
Transfer Auth users from Firebase

### DISCUSSION THREAD

**Discussion by cdedreuille** [supabase/supabase#1264] [Questions] [answered]
I was wondering if anyone successfully transfers all auth users from Firebase to Supabase? We have around 1k users and I'm trying to find the best way to move them to Supabase. Did someone made this in the past?

**Comment by silentworks** [comment]
You could create a script to grab all of them from Firebase and put them in an array, then use a postgres library to insert them using a transaction. Or you could probably export them as CSV from Firebase and then use an external SQL application to import the CSV into the Postgres database on Supabase.

**Reply by cdedreuille** [reply]
Yes, that's pretty much what I'll do, but I wondered if someone already did it and by chance documented the process. Since Supabase is defining themselves as a Firebase alternative, I believe this will be quite a common use case.

**Reply by silentworks** [reply]
Oh right. This might make a great article or video for Supabase. If I get some time I will try and write something up on how I would approach this with code examples.

**Reply by heauton** [reply]
Hey @silentworks I'm not sure I follow. How do we handle passwords for existing users then? I was able to export firebase users using https://firebase.google.com/docs/cli/auth#auth-export but I have no clue how to import them into supabase users so they can keep their current passwords. Could you elaborate a little more? Did @cdedreuille get a chance to transfer users?

**Reply by cdedreuille** [reply]
So I did succeed to import users but not passwords. I added a column on our users' table to check if this user needs to be converted to Supabase. Then our login page, after they put their email and password, I check if the user exist in Supabase and if it needs to be converted, then I create a new auth user in Supabase, take the new ID and update it in the user table and other places.

It's obviously a bit hacky but that way I'm not forcing the user to create a new password.

**Reply by marcelscruz** [reply]
This seems to be a possible workaround indeed, @cdedreuille, just let the users do the transition by themselves.

The one issue I see here is that the old DB needs to stay active for as long as all users did the migration, which might never happen. Do I get it right?

I was considering a way of adding all users into the new DB and take the new ID to map them to other related tables (a bit of a hack here). With that, it's like every user is "Waiting for verification", just like when you use the "Invite" button.
Next time the user comes to the platform, they'll have to either use external OAuth (would only feel like they lost their auth session) or redefine the password through a "reset password email" (not ideal 🤷‍♂️).

Any thoughts on that?

By the way, is there a way to bulk add users to the DB, through the SQL editor maybe? Doing one by one through the "Invite" button will take a little while 😬.

**Reply by cdedreuille** [reply]
In my case, I imported all users through SQL I believe or using a javascript function (I'm using Nextjs API). So no need for the old DB anymore. When they login and the auth user doesn't exist, then I create a new auth user and replace the ID in my user table.

**Reply by marcelscruz** [reply]
I see, that makes sense. I'll try to implement something similar then. Thank you for the tip!

**Reply by liaujianjie** [reply]
Not sure if this will be of interest to you folks but I thought I'll leave a note in case anyone need this.

I managed to migrate my users from Firebase Auth to Supabase GoTrue by inserting directly into the `users` table on the `auth` schema. This is how I did it:
1. Export Firebase Auth data via `firebase auth:export ...`
2. Write a node.js script that:
   1. Reads from that exported JSON from the previous step
   2. Validates the JSON (more on this below)
   3. Map the JSON data into insert statements for the `auth.users` table
   4. Batch inserts 10k users at a time in a transaction block with `pg-promise`
   5. You should observe that your users are now in the Supabase dashboard.

Some notes:
- Unfortunately, this only works for users that do not use passwords. I was lucky enough that I never allowed password logins on my site.
- Validating the JSON is important because it seems GoTrue "merges" logins via different 3rd part auth providers via email. That is, the email address of the user uniquely identifies the user. As such, you need to make sure that all your users have an email address. I'm not super familiar with GoTrue so maybe someone from the Supabase team can verify this.
  - It is possible for your Firebase Auth user to not have an email address if you did not provide enough scopes when your user logged in on firebase. For example, using Facebook login with Firebase Auth out of the box doesn't provide you with the email.

I would appreciate it if the Supabase team can comment on the correctness of my migration. 😅

**Reply by marcelscruz** [reply]
Thank you for sharing this @liaujianjie!

I believe that for users with password I'll have to add a check on their login attempt and ask them to redefine it, in case they didn't yet.

About the script you wrote, would it be possible to share it on a Gist or something? I'm not used to writing SQL statements, so not sure how to go about it.

Thanks!

**Reply by liaujianjie** [reply]
I'll make a public repo for it sometime this week. :-)

Done, see here: https://github.com/supabase/supabase/discussions/1264#discussioncomment-1093927

**Comment by liaujianjie** [comment, marked as answer]
I was too caught up at work over the past few weeks so I didn't have time to put it out until today. But anyways here's the migration repo [as promised](https://github.com/supabase/supabase/discussions/1264#discussioncomment-900620): https://github.com/liaujianjie/firebase-to-supabase-auth-migrator

I've not tested it because I made it by pulling code out of my team's internal migrator scripts. Happy to answer any questions or accept PRs if you folks have any.

**Reply by k0shk0sh** [reply]
unfortunately this doesn't work  as confirmed_at is auto generated colum.

```
length: 158,
  severity: 'ERROR',
  code: '42601',
  detail: 'Column "confirmed_at" is a generated column.',
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'rewriteHandler.c',
  line: '829',
  routine: 'rewriteTargetListIU'
```

**Reply by k0shk0sh** [reply]
I fixed the errors and opened up a PR 

https://github.com/liaujianjie/firebase-to-supabase-auth-migrator/pull/1
